### PR TITLE
KEYCLOAK-17615 Move database initialization to JpaConnectionProviderFactory

### PR DESCRIPTION
**Pull request by vramik**
`migrateModel(KeycloakSession)` is executed conditionally. When all realm, client and clientScopes providers are `jpa`. It's based on assumtion that in map store we can do it without migration of model.

### CODE REVIEW

File: model/map/src/main/java/org/keycloak/models/map/serverinfo/MapServerInfoProviderFactory.java [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+        }
+    }
+
+    @Override
+    public void postInit(KeycloakSessionFactory factory) {
+    }
+
+    @Override
+    public ServerInfoProvider create(KeycloakSession session) {
+        return new ServerInfoProvider() {
```
Comment (hmlnarik) [inline model/map/src/main/java/org/keycloak/models/map/serverinfo/MapServerInfoProviderFactory.java:59]: Can either `MapServerInfoProviderFactory` just also implement `ServerInfoProvider` and here return `this`, or even better, prepare this `ServerInfoProvider` as a `private static final` instance and return it?

**Review by hmlnarik — COMMENTED** [review]
Thanks @vramik, overall looks good to me, please see inline just few intermediate comments

File: model/map/src/main/java/org/keycloak/models/map/serverinfo/MapServerInfoProviderFactory.java [hunk 2]
```diff
@@ -26,18 +26,19 @@
...
 import org.keycloak.models.ServerInfoProviderFactory;
+import org.keycloak.models.map.common.AbstractMapProviderFactory;
 
-public class MapServerInfoProviderFactory implements ServerInfoProviderFactory {
+public class MapServerInfoProviderFactory extends AbstractMapProviderFactory<ServerInfoProvider> implements ServerInfoProviderFactory {
 
     @Override
     public void init(Config.Scope config) {
-        String seed = config.get("resourcesVersionSeed");
+        String seed = config.get(Constants.RESOURCES_VERSION_SEED);
```
Comment (hmlnarik) [inline model/map/src/main/java/org/keycloak/models/map/serverinfo/MapServerInfoProviderFactory.java:39]: Please have this constant local to `MapServerInfoProviderFactory` class.

**Comment by vramik** [comment]
With current implementation `Version.RESOURCES_VERSION` will be correctly set only in case all `realm`, `client` and `clientScope` providers are `jpa` or `serverInfo` provider is `map`. I'm not sure if it is something we want.

File: quarkus/runtime/src/main/java/org/keycloak/connections/jpa/QuarkusJpaConnectionProviderFactory.java [hunk 3]
```diff
@@ -240,9 +240,9 @@ public void run(KeycloakSession session) {
                      * Migrate model is executed just in case following providers are "jpa".
                      * In Map Storage, there is an assumption that migrateModel is not needed.
                      */
-                    if ("jpa".equals(Config.getProvider("realm")) &&
-                        "jpa".equals(Config.getProvider("client")) &&
-                        "jpa".equals(Config.getProvider("clientScope"))) {
+                    if (Config.getProvider("realm") != null && "jpa".equals(Config.getProvider("realm")) &&
```
Comment (hmlnarik) [inline quarkus/runtime/src/main/java/org/keycloak/connections/jpa/QuarkusJpaConnectionProviderFactory.java:243]: `Config.getProvider("realm") != null` effectively means `Config.getProvider("realm") == "jpa"`, as `jpa` is (currently) a default provider.

Comment (vramik) [reply quarkus/runtime/src/main/java/org/keycloak/connections/jpa/QuarkusJpaConnectionProviderFactory.java:243]: `Config.getProvider("realm")` here returns null. There are links to code where it's resolved: 

- https://github.com/keycloak/keycloak/blob/master/quarkus/runtime/src/main/java/org/keycloak/configuration/MicroProfileConfigProvider.java#L44
- https://github.com/keycloak/keycloak/blob/master/quarkus/runtime/src/main/java/org/keycloak/configuration/MicroProfileConfigProvider.java#L64

So in order to make tests pass I've added the null check, but maybe it's not correct approach.

File: server-spi-private/src/main/java/org/keycloak/migration/MigrationModelManager.java [hunk 4]
```diff
@@ -100,7 +100,7 @@
     public static void migrate(KeycloakSession session) {
         session.setAttribute(Constants.STORAGE_BATCH_ENABLED, Boolean.getBoolean("keycloak.migration.batch-enabled"));
         session.setAttribute(Constants.STORAGE_BATCH_SIZE, Integer.getInteger("keycloak.migration.batch-size"));
-        MigrationModel model = session.getProvider(ServerInfoProvider.class).getMigrationModel();
+        MigrationModel model = session.getProvider(ServerInfoProvider.class, "jpa").getMigrationModel();
```
Comment (hmlnarik) [inline server-spi-private/src/main/java/org/keycloak/migration/MigrationModelManager.java:103]: Why is this change here?

Comment (vramik) [reply server-spi-private/src/main/java/org/keycloak/migration/MigrationModelManager.java]: I forgot to revert this. It's from my local tests, reverting, thank you.

**Comment by vramik** [comment]
https://keycloak-jenkins.com/job/universal-test-pipeline-adapters/203/
https://keycloak-jenkins.com/job/universal-test-pipeline-server/315/

File: model/map/src/main/java/org/keycloak/models/map/serverinfo/MapServerInfoProviderFactory.java [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+import org.keycloak.migration.MigrationModel;
+import org.keycloak.migration.ModelVersion;
+import org.keycloak.models.KeycloakSession;
+import org.keycloak.models.ServerInfoProvider;
+import org.keycloak.models.ServerInfoProviderFactory;
+import org.keycloak.models.map.common.AbstractMapProviderFactory;
+
+public class MapServerInfoProviderFactory extends AbstractMapProviderFactory<ServerInfoProvider> implements ServerInfoProviderFactory {
+
+    private static final String RESOURCES_VERSION_SEED = "resourcesVersionSeed";
```
Comment (mhajas) [inline model/map/src/main/java/org/keycloak/models/map/serverinfo/MapServerInfoProviderFactory.java:36]: What is the purpose of this seed?

**Review by mhajas — DISMISSED** [review]
Thank you @vramik. The PR looks good to me,  I just added one question because I wasn't sure what it is for.

File: model/map/src/main/java/org/keycloak/models/map/serverinfo/MapServerInfoProviderFactory.java (hunk 5 above)
Comment (vramik) [reply model/map/src/main/java/org/keycloak/models/map/serverinfo/MapServerInfoProviderFactory.java:36]: It is useful in clustered scenarios. When this seed is not set in config, the random seed is generated and used. In clustered environment it might be useful to set the seed yourself for having the same `Version.RESOURCES_VERSION` across all nodes in cluster. Then all resources (js, css, ...) could have the same url.

Comment (mhajas) [reply model/map/src/main/java/org/keycloak/models/map/serverinfo/MapServerInfoProviderFactory.java:36]: I see. Ok thanks for the explanation.

**Review by mhajas — DISMISSED** [review]

**Review by hmlnarik — APPROVED** [review]
Thanks for the PR!
